File: Assignment 1/Codes/Assignment1.cpp
#include <bits/stdc++.h>
#include <random>

using namespace std;

double random()
{
    random_device rd;
    mt19937 generator(rd());
    uniform_real_distribution<double> u(0.0, 1.0);
    return u(generator);
}

int main()
{
    double R=1+random();
    double L = 20*R;
    int Nmax = pow(L,3)*0.74*3/(4*3.14*pow(R,3));
    int M = 100;
    vector<int> N(M);
    vector<double> rmean(M);

    for(int i=0; i<M; i++)
    {
        int accept=0;
        int reject=0;
        vector<vector<double>> spheres{{R + (L-2*R)*random(),R + (L-2*R)*random(),R + (L-2*R)*random()}};
        while (reject<10*Nmax)
        {
            int tmp=0;
            double rm=0;
            double x1 = R + (L-2*R)*random();
            double y1 = R + (L-2*R)*random();
            double z1 = R + (L-2*R)*random();
            for(int j=0; j<spheres.size(); j++)
            {
                double r = sqrt(pow((x1-spheres[j][0]),2)+pow((y1-spheres[j][1]),2)+pow((z1-spheres[j][2]),2));
                if(r>=2*R)
                {
                    rm = rm+r;
                }
                else
                {
                    tmp=1;
                    break;
                }
            }
            if(tmp==1)
            {
                reject++;
            }
            else
            {
                reject=0;
                accept++;
                rmean[i] = rmean[i]+rm;
                spheres.push_back({x1,y1,z1});

            }
        }
        N[i] = accept;
        rmean[i] = 2*rmean[i]/(N[i]*(N[i]-1));
        cout << i << endl;
    }
    return 0;
}

File: Assignment 1/Codes/Assignment 1.py
import numpy as np
import random
import math as m
import matplotlib.pyplot as plt
from itertools import product, combinations

R = 1+random.random()
L = 20*R

Nmax = int(L**3*0.74*3/(4*m.pi*R**3))
#print(R,L,Nmax)

M = 1
N = np.zeros((M))
rmean = np.zeros((M))

for i in range(M):
    reject=0
    accept=0
    spheres = [[R+(L-2*R)*random.random(),R+(L-2*R)*random.random(),R+(L-2*R)*random.random()]]

    while(reject<10*Nmax):
        tmp=0
        rm = 0
        x1 = R+(L-2*R)*random.random()
        y1 = R+(L-2*R)*random.random()
        z1 = R+(L-2*R)*random.random()
        for sphere in spheres:
            r = m.sqrt((x1-sphere[0])**2+(y1-sphere[1])**2+(z1-sphere[2])**2)
            if(r>=2*R):
                rm = rm+r
            else:
                tmp = 1
                break
        if(tmp):
            reject = reject+1
        else:
            reject=0
            accept = accept+1
            rmean[i] = rmean[i]+rm
            spheres = spheres+[[x1,y1,z1]]
    N[i] = accept
    rmean[i] = 2*rmean[i]/(N[i]*(N[i]-1))

print(N)
print(rmean)

fig = plt.figure()


ax = fig.add_subplot(projection='3d')
r = [0, L]
for s, e in combinations(np.array(list(product(r, r, r))), 2):
   if np.sum(np.abs(s-e)) == r[1]-r[0]:
      ax.plot3D(*zip(s, e), color="black")
for sphere in spheres:
    ax.scatter(sphere[0], sphere[1], sphere[2], c='y', s=10)

plt.show()